docker/internal/tarfile/src.go

```diff
@@ -66,43 +70,67 @@ func (s *Source) ensureCachedDataIsPresent() error {
 // Call ensureCachedDataIsPresent instead.
 func (s *Source) ensureCachedDataIsPresentPrivate() error {
 	// Read and parse manifest.json
-	tarManifest, err := s.loadTarManifest()
+	manifests, err := s.loadTarManifest()
 	if err != nil {
 		return err
 	}
 
-	// Check to make sure length is 1
-	if len(tarManifest) != 1 {
-		return errors.Errorf("Unexpected tar manifest.json: expected 1 item, got %d", len(tarManifest))
+	tarManifest, err := chooseManifest(manifests, s.ref)
+	if err != nil {
+		return err
 	}
 
 	// Read and parse config.
-	configBytes, err := s.archive.readTarComponent(tarManifest[0].Config, iolimits.MaxConfigBodySize)
+	configBytes, err := s.archive.readTarComponent(tarManifest.Config, iolimits.MaxConfigBodySize)
 	if err != nil {
 		return err
 	}
 	var parsedConfig manifest.Schema2Image // There's a lot of info there, but we only really care about layer DiffIDs.
 	if err := json.Unmarshal(configBytes, &parsedConfig); err != nil {
-		return errors.Wrapf(err, "Error decoding tar config %s", tarManifest[0].Config)
+		return errors.Wrapf(err, "Error decoding tar config %s", tarManifest.Config)
 	}
 	if parsedConfig.RootFS == nil {
-		return errors.Errorf("Invalid image config (rootFS is not set): %s", tarManifest[0].Config)
+		return errors.Errorf("Invalid image config (rootFS is not set): %s", tarManifest.Config)
 	}
 
-	knownLayers, err := s.prepareLayerData(&tarManifest[0], &parsedConfig)
+	knownLayers, err := s.prepareLayerData(tarManifest, &parsedConfig)
 	if err != nil {
 		return err
 	}
 
 	// Success; commit.
-	s.tarManifest = &tarManifest[0]
+	s.tarManifest = tarManifest
 	s.configBytes = configBytes
 	s.configDigest = digest.FromBytes(configBytes)
 	s.orderedDiffIDList = parsedConfig.RootFS.DiffIDs
 	s.knownLayers = knownLayers
 	return nil
 }
 
+// chooseManifest selects an item from manifests matching ref, which may be nil to mean the only image.
+func chooseManifest(manifests []ManifestItem, ref reference.NamedTagged) (*ManifestItem, error) {
```

Comment:
I think we also need a way to chose _by index_ (or more complex by digest). I created the following `manifest.json` with `docker-save` with two untagged images.
```json
[
  {
    "Config": "1c35c441208254cb7c3844ba95a96485388cef9ccc0646d562c7fc026e04c807.json",
    "RepoTags": null,
    "Layers": [
      "9719c1ce1e2335e5b108c339b9f2b2f8aae8e62f30a831c40b9d9329678fd711/layer.tar"
    ]
  },
  {
    "Config": "a24bb4013296f61e89ba57005a7b3e52274d8edd3ae2077d04395f806b63d83e.json",
    "RepoTags": null,
    "Layers": [
      "fa27ccac46e9a5b7ff1d188e99763db1a5fb45adb34a917c1c482dfde99ad432/layer.tar"
    ]
  }
]
```

The RepoTags of both manifests is "null", so we cannot match by reference.

Maybe that's being tackled in a later commit. Will read further.